clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/KafkaDataIntegrityValidator.java

```diff
@@ -97,4 +130,37 @@ public void checkMissingMessage(
     producerTracker.checkMissingMessage(consumerRecord, errorMetricCallback, this.kafkaLogCompactionDelayInMs);
   }
 
+  void clearExpiredState() {
+    if (this.maxAgeInMs != DISABLED) {
+      long currentTime = time.getMilliseconds();
+      long maxRecordProducerTimestamp = currentTime - this.maxAgeInMs;
+      if (this.oldestRetainedProducerTS < maxRecordProducerTimestamp) {
+        synchronized (this) {
+          if (this.oldestRetainedProducerTS < maxRecordProducerTimestamp) {
+            int numberOfClearedGUIDs = 0;
+            long newOldestRetainedProducerTS = Long.MAX_VALUE;
+            Iterator<ProducerTracker> iterator = this.producerTrackerMap.values().iterator();
+            ProducerTracker producerTracker;
+            while (iterator.hasNext()) {
+              producerTracker = iterator.next();
+              long currentProducerOldestRetainedTS = producerTracker.clearExpiredState(maxRecordProducerTimestamp);
+              if (currentProducerOldestRetainedTS == Long.MAX_VALUE) {
+                iterator.remove();
+                numberOfClearedGUIDs++;
+              } else {
+                newOldestRetainedProducerTS = Math.min(newOldestRetainedProducerTS, currentProducerOldestRetainedTS);
+              }
+            }
+            if (newOldestRetainedProducerTS != Long.MAX_VALUE) {
```

Comment:
If there is no state retained, shall we update `newOldestRetainedProducerTS` to the current time to avoid the empty while loop?
Maybe it doesn't matter since empty loop is cheap..

Reply:
I don't use this field anymore in the latest iteration. Maybe it could be brought back in some way if the state clearing algorithm is feared to be too expensive, but I'm not sure ATM.